### DISCUSSION TITLE
Problem about print cosseratInSofaFrameNode.FramesMO.velocity

### DISCUSSION THREAD

**Discussion by xlsufo** [sofa-framework/sofa#4966] [Cosserat] [answered]
Hello,
I want to obtain the position and velocity of nodes in a cosserat rod. I used the following code to obtain them.
![Quicker_20240903_223000](https://github.com/user-attachments/assets/a34e4451-5ac2-410e-a1eb-2658dfe2cd80)

But only the position result is correct, and the number of velocity matrices does not correspond to the number of nodes.
![Quicker_20240903_222910](https://github.com/user-attachments/assets/040a1a1c-1c26-4e2e-8489-ed1bdf271272)

The number of nodes I set is 17, and the SOFA version is v23.06. I am very confused, may I ask how to solve it？

**Comment by bakpaul** [comment]
Hello ! 

I am sorry but the version you are using is a quite old one. I know that a lot of issue have been fixed since this release, the plugin has changed a lot (for the better). 

I can notify the core developer of this plugin : @adagolodjo to see if he has any idea on your issue. 

But finally, I think that the best thing to do is to update your version of SOFA, because if there is an issue in your version of Cosserat, it has either already been corrected in the newest version, or it will be corrected there, but not on this branch...

Hope this helped you.

**Reply by xlsufo** [reply]
Thank you very much, I will try it in the new version as soon as possible to see if the problem still exists.

**Reply by xlsufo** [reply]
Hello！
I have tried the new version of SOFA, but when I run the example in the cosserat plugin, the following error always appears
![屏幕截图 2024-09-07 233136](https://github.com/user-attachments/assets/7d382d3b-8198-4666-baac-3aed4a74ced4)

**Reply by bakpaul** [reply]
Can you tell me what is the example you are running ?

**Reply by xlsufo** [reply]
PCS_Example1.py, PCS_Example2.py and PNLS_Example1.py.

**Comment by adagolodjo** [comment]
Hello @xlsufo,

Please follow the tutorial scenes created in this folder "https://github.com/SofaDefrost/Cosserat/tree/master/tutorial/tuto_scenes". To understand how to create Cosserat scenes step by step.

For the problem with the size of the velocity array. Can you tell us if you have a good size of the velocity array in the graph mode ?
Like => ![image](https://github.com/user-attachments/assets/599494c0-c129-4c88-a67b-ee26a1f932ea)

**Comment by xlsufo** [comment]
Thank you for your patient answer, but for this new version, I just opened the example PCS_Example1.py. Is the following the picture you want to see
![屏幕截图 2024-09-11 222531](https://github.com/user-attachments/assets/c09cbed2-c5ca-410b-8ebd-afdda90cecb3)

**Comment by adagolodjo** [comment]
Yes, @xlsufo. 

This is the speed you are trying to print. And as you can see here, you have more than 3 values. Although it is zero here, I guess you haven't run your scene yet.

**Reply by xlsufo** [reply]
Thank you very much, @adagolodjo.
In my original question, I started the SOFA GUI via python, and everything else worked fine, except the velocity output.
There is the key code: 
``` python 
     allPose=solverNode.needle.rigidBase.cosseratInSofaFrameNode.FramesMO.position.array()
    allVec = solverNode.needle.rigidBase.cosseratInSofaFrameNode.FramesMO.velocity.array()
    solverNode.needle.rigidBase.cosseratInSofaFrameNode.addObject(FingerController(allPose,allVec))
```

```python
class FingerController(Sofa.Core.Controller):
    def __init__(self, *args, **kwargs):
        Sofa.Core.Controller.__init__(self, *args, **kwargs)
        self.allPos = args[0]
        self.allVec = args[1]
        self.name = "FingerController"
    def onEvent(self, params):        
        posi = self.allPos
        print('posi',posi)
        vec = self.allVec
        print('vec',vec)
```
Here is the output:

```
posi [[-1.17835932e-08 -1.18374134e-27  1.55277113e-08  3.86101022e-30
  -2.22181382e-09 -6.82348958e-29  1.00000000e+00]
 [ 2.49999674e-02 -9.94173112e-25  2.79809017e-05  2.44887300e-23
  -1.11861297e-03 -3.97012526e-23  9.99999374e-01]
 [ 4.99998214e-02 -3.97495566e-24  1.11876661e-04  4.89774254e-23
  -2.23722232e-03 -7.94023873e-23  9.99997497e-01]
 [ 7.49995237e-02 -8.20194335e-24  2.33748798e-04  2.82434064e-23
  -2.63767768e-03 -8.94907848e-23  9.99996521e-01]
 [ 9.99991203e-02 -1.29311751e-23  3.75643483e-04  7.50938286e-24
  -3.03813262e-03 -9.95791680e-23  9.99995385e-01]
 [ 1.24998663e-01 -1.74822833e-23  5.26879220e-04 -2.81080693e-24
  -3.01132451e-03 -8.24517352e-23  9.99995466e-01]
 [ 1.49998213e-01 -2.11754740e-23  6.76774569e-04 -1.31309967e-23
  -2.98451641e-03 -6.53243024e-23  9.99995546e-01]
 [ 1.74997800e-01 -2.16022819e-23  8.20486414e-04 -1.61686405e-23
  -2.76398116e-03  4.81678017e-23  9.99996180e-01]
 [ 1.99997448e-01 -1.63542633e-23  9.53171620e-04 -1.92062835e-23
  -2.54344577e-03  1.61659903e-22  9.99996765e-01]
 [ 2.24997160e-01  4.37430190e-24  1.07313107e-03 -2.04221211e-23
  -2.25494607e-03  6.67390013e-22  9.99997458e-01]
 [ 2.49996937e-01  5.03891708e-23  1.17866564e-03 -2.16379571e-23
  -1.96644618e-03  1.17312007e-21  9.99998067e-01]
 [ 2.74996772e-01  1.54479889e-22  1.26922244e-03 -2.33387165e-23
  -1.65583170e-03  2.99043392e-21  9.99998629e-01]
 [ 2.99996659e-01  3.49436021e-22  1.34424857e-03 -2.50394737e-23
  -1.34521705e-03  4.80774749e-21  9.99999095e-01]
 [ 3.24996589e-01  7.23459422e-22  1.40337077e-03 -2.79770603e-23
  -1.01967245e-03  1.01531361e-20  9.99999480e-01]
 [ 3.49996552e-01  1.36475194e-21  1.44621576e-03 -3.09146439e-23
  -6.94127746e-04  1.54985236e-20  9.99999759e-01]
 [ 3.74996538e-01  2.46518070e-21  1.47226163e-03 -3.72522289e-23
  -3.47707278e-04  2.85185971e-20  9.99999940e-01]
 [ 3.99996536e-01  4.21661264e-21  1.48098648e-03 -4.35898093e-23
  -1.28676939e-06  4.15386671e-20  1.00000000e+00]]
vec [[2.12199579e-314 1.46484787e-311 1.46438711e-311 1.46484792e-311
  6.95260666e-310 1.46484787e-311]]
```
The GUI displays the following content:
![Quicker_20240912_092518](https://github.com/user-attachments/assets/8b942357-75ba-439a-b84f-a2086271fbfa)

**Reply by adagolodjo** [reply, marked as answer]
Hi @xlsufo,

Base on by PCS_Example1.py

       `position = self.frames.position.value
        print(f' ==>The position of the beam tip is : {position}')
        velocity = self.frames.velocity.value
        print(f' ==>The velocity of the beam tip is : {velocity}')
        `
This should work. Change `array()` by `value`

**Reply by xlsufo** [reply]
Hi @adagolodjo，
Thank you very much. All the problems have been solved. I have seen the latest tutorial and I will try to use the CosseratBase class in future projects.

**Comment by adagolodjo** [comment]
@xlsufo, I also fixed the issue related to the error you mentioned in your second post, you can find the PR here => https://github.com/SofaDefrost/Cosserat/pull/130. 

This will be merged to Cosserat's master soon.
